lgc/builder/BuilderImpl.cpp

```diff
@@ -172,16 +173,63 @@ Value *BuilderImplBase::CreateIntegerDotProduct(Value *vector1, Value *vector2,
           indices = ArrayRef<int>({2, 3});
           input1 = CreateShuffleVector(input1, Constant::getNullValue(inputTy), indices);
           input2 = CreateShuffleVector(input2, Constant::getNullValue(inputTy), indices);
-          product =
+          scalar =
               CreateIntrinsic(intrinsicDot2, {}, {input1, input2, intermediateRes, getInt1(false)}, nullptr, instName);
         }
         // Add a satruation add if required.
-        if (!hasNoAccumulator)
-          scalar = CreateNamedCall("llvm.sadd.sat.i32", getInt32Ty(), {product, accumulator}, {}, instName);
+        if (hasAccumulator)
+          scalar = CreateNamedCall("llvm.sadd.sat.i32", getInt32Ty(), {scalar, accumulator}, {}, instName);
       }
     }
   }
 
+  // Do clamp if it has an accumulator
+  // NOTE: HW supported dot product will do clamp for 32-bit accumulator and no need to do it here
+  if ((!isHwSupported || (isHwSupported && scalar->getType() != outputTy))) {
+    if (hasAccumulator) {
+      // The scope for 8-bit/32-bit/16-bit/64-bit signed/unsinged integer type
```

Comment:
Can we just calculate these constants from bitWidth?

```
unsignedMax = (2ULL << (bitWidth - 1)) - 1;
signedMax = unsignedMax >> 1;
signedMin = -1ULL - signedMax;
```

Also I think this code does nothing for 64-bit, so that case can be omitted.

Are we expecting to get the correct results for 64-bit vector inputs that saturate? I think that will give an incorrect result.